neqo-transport/src/recovery/mod.rs

```diff
@@ -828,7 +828,12 @@ impl LossRecovery {
     /// When it has, mark a few packets as "lost" for the purposes of having frames
     /// regenerated in subsequent packets.  The packets aren't truly lost, so
     /// we have to clone the `SentPacket` instance.
-    fn maybe_fire_pto(&mut self, rtt: &RttEstimate, now: Instant, lost: &mut Vec<SentPacket>) {
+    fn maybe_fire_pto(&mut self, primary_path: &PathRef, now: Instant, lost: &mut Vec<SentPacket>) {
+        let path = primary_path.borrow();
+        if !path.is_valid() {
+            return;
```

Comment:
Sorry, I just realized that the root cause of this PR is not the same as #1491 .
That assertion `earliest > now` is caused by this line. If I remove the early return, the test `idle_timeout_crazy_rtt` passes.

I am still looking into why this change can trigger that assertion.

Reply:
@KershawChang any insights?

Reply:
@KershawChang ping?

Reply:
I just tried this (i.e., removing the early return) and while that satisfies the `earliest > now` assertion, the `loss_time_past_largest_acked` test now fails:
```
thread 'connection::tests::recovery::loss_time_past_largest_acked' panicked at neqo-transport/src/connection/tests/recovery.rs:587:5:
assertion `left != right` failed
  left: 0ns
 right: 0ns
stack backtrace:
   0: rust_begin_unwind
             at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/std/src/panicking.rs:652:5
   1: core::panicking::panic_fmt
             at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/core/src/panicking.rs:72:14
   2: core::panicking::assert_failed_inner
   3: core::panicking::assert_failed::<core::time::Duration, core::time::Duration>
             at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/core/src/panicking.rs:364:5
   4: neqo_transport::connection::tests::recovery::loss_time_past_largest_acked
             at ./src/connection/tests/recovery.rs:587:5
   5: neqo_transport::connection::tests::recovery::loss_time_past_largest_acked::{closure#0}
             at ./src/connection/tests/recovery.rs:560:34
   6: <neqo_transport::connection::tests::recovery::loss_time_past_largest_acked::{closure#0} as core::ops::function::FnOnce<()>>::call_once
             at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/core/src/ops/function.rs:250:5
   7: core::ops::function::FnOnce::call_once
             at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/core/src/ops/function.rs:250:5
             ```

Reply:
Sorry, I haven't made much progress. I'll try to look at this again this week.

Reply:
I think I have a fix for the test (similar to how I needed to fix another test in this PR).